### DISCUSSION TITLE
How to use it with React Three Fiber?

### DISCUSSION THREAD

**Discussion by dulre159** [creativelifeform/three-nebula#181] [Q&A] [answered]
Hi i would like to use this with react three fiber: https://github.com/pmndrs/react-three-fiber
It is possibile?
Are there examples to follow?
I am trying to create an explosion like animation. 
Thanks :)

**Comment by rohan-deshpande** [comment, marked as answer]
Hi there, right now there's no way to easily use `three-nebula` with `react-three-fiber` however I am planning on writing a library to make this work.

**Reply by dulre159** [reply]
Great! Thanks you!

**Reply by BennyKok** [reply]
Was wondering if doing something like this can temporality work?

Like if we can get the scene reference like this, 

```js
const threeScene = useRef()
```

```js
 useEffect(() => {
    const renderer = new SpriteRenderer(threeScene, THREE);
    System.fromJSONAsync(TestLaserEffect, THREE).then(system => {
      console.log(system);
      system.addRenderer(renderer)
      system.emit(...)
    });
  }, [])
```

```js
  return (
    <Canvas {...props} >
      <scene ref={threeScene}>
      </scene>
    </Canvas>
   )
```

**Reply by rohan-deshpande** [reply]
Might work, I've never used `react-three-fibre`, but definitely try this out and see how it goes

**Reply by ThorstenBux** [reply]
Hi there, thanks for the question and the proposed answer. For who ever it might help. (I just started dabbling with nebula). I've integrated it into R3F like this:
```
const nebula = useRef<any>(null);
const {scene: threeScene} = useThree()

    useFrame(({scene, camera, gl: renderer}, delta) => {
        if (!nebula.current) return
        // This function runs 60 times/second inside the global render-loop
        nebula.current.update();
        renderer.render(scene, camera);
      })



      useEffect(() => {
        Nebula.fromJSONAsync(json, THREE).then((loaded: any) => {
            const nebulaRenderer = new SpriteRenderer(threeScene, THREE);
            nebula.current = loaded.addRenderer(nebulaRenderer);
          
          });
      }, [threeScene])
```

**Reply by code-matt** [reply]
You can also just render a `<group ref={ref} />` and pass that ref instead of scene in order to attach the emitter to your component

**Comment by Alexp120** [comment]
I converted Thorstens example, however, if your anything like me it didn't work with the nebula json generator, the desktop program for creating the particle animations i mean.

This is because when you load the json it also loads all of the data instead of just the portion you need and it leads to no data being transferred to the nebulaRenderer, so here is the code I made for my smoke effect.

```
import { useRef, useEffect } from "react";
import { useThree, useFrame } from "@react-three/fiber";
import * as THREE from "three";
import Nebula, { SpriteRenderer } from "three-nebula";
import json from "./effects/smoke-effect.json"; // my path to the smoke effect json data

const SmokeEffect = () => {
const nebula = useRef(null);
const { scene: threeScene } = useThree();

useFrame(({ gl, scene, camera }) => {
if (nebula.current) {
nebula.current.update();
gl.render(scene, camera);
}
});

useEffect(() => {
Nebula.fromJSONAsync(json.particleSystemState, THREE).then((loaded) => { // if you just have json here it wont display anything
const nebulaRenderer = new SpriteRenderer(threeScene, THREE);
nebula.current = loaded.addRenderer(nebulaRenderer);
});
}, [threeScene]);

return null;
};

export default SmokeEffect;

```
After you add this make sure to import your effect name that you exported and include that in your canvas to make it render.
 I personally like this far more than regular 3js because it's much more concise and readable, other than a bit of a headache on setup.

im adding this because this solution does not exist on the internet and I wish it did, I spent 4 hours figuring this out so enjoy!

also if you want to re-position where the effect is you have to do that either before it generates the json or within the position property of the json itself, to my knowledge. Im working on a custom positioning method but this is my current work around.
